### DISCUSSION TITLE
Feedback needed on new UI - Axis & Unit Interoperability

### DISCUSSION THREAD

**Discussion by scottrenaud** [Autodesk/maya-usd#3728] [Show and tell]
Hey folks,

We’d like your feedback on some designs for Axis & Unit Interoperability, where we're looking to improve support of USDs stage metadata **upAxis** and **metersPerUnit**, so that things look good when artists bring things in.

![axis_unit_banner](https://github.com/Autodesk/maya-usd/assets/108303538/d138983f-3da0-497a-99f1-d6fe445c5c23)

<h2>Phase 1 - Import, Export, Create stage from file</h2>
<ul><li>When importing a USD file or creating a stage from file, based on user settings:</li>
<ul><li>read the stage metadata (upAxis/metersPerUnit)</li>
<li>determine if it matches your scene settings or not</li>
<li>perform a conversion so that it looks right (if necessary)</li></ul>
 <li>Give the flexibility to determine how the conversion should be carried out:</li>
<ul><li>Rotate and/or Scale</li>
 <li>Add Parent Transform</li>
 <li>Overwrite Maya Preferences</li></ul>
 <li>When exporting to USD, allow the artist to convert to a different upAxis or unit and write the stage metadata (upAxis/metersPerUnit)</li>
 <li>Axis will be converted via 90 degree rotation</li>
 <li>Units will be converted via a scaling factor applied to the scale</li></ul>

<h3>Watch below!</h3>

**Axis/Unit conversion on Import - Rotate and/or Scale**
<video src="https://github.com/Autodesk/maya-usd/assets/108303538/f7f81115-1262-49c1-8bf7-cb3006446ece" controls="controls" muted="muted" class="d-block rounded-bottom-2 width-fit" style="max-height:640px;">

**Axis/Unit correction on Import - Add Parent Transform**
<video src="https://github.com/Autodesk/maya-usd/assets/108303538/8a18904f-20dc-4bb9-a740-b68cac4ee8e2" controls="controls" muted="muted" class="d-block rounded-bottom-2 width-fit" style="max-height:640px;">

**Axis/Unit correction on Import - Overwrite Maya Preferences**
<video src="https://github.com/Autodesk/maya-usd/assets/108303538/591c3310-2400-48af-abd2-75ceb2b55589" controls="controls" muted="muted" class="d-block rounded-bottom-2 width-fit" style="max-height:640px;">

**Export - Use Maya Preferences**
<video src="https://github.com/Autodesk/maya-usd/assets/108303538/e746b7a0-8d72-4d05-a139-ce068595a56f" controls="controls" muted="muted" class="d-block rounded-bottom-2 width-fit" style="max-height:640px;">

**Create Stage from File -  Rotate and/or Scale**
<video src="https://github.com/Autodesk/maya-usd/assets/108303538/37012a8f-8b97-45b1-8571-54a8a012c3b6" controls="controls" muted="muted" class="d-block rounded-bottom-2 width-fit" style="max-height:640px;">

<h2>Beyond</h2>

_Phase 2_ -  Support other basic workflows (duplicate, edit as maya, cache to usd, USD references, etc.)
_Phase 3_  - Authoring directly in USD (note there's currently a [proposal](https://github.com/spiffmon/OpenUSD-proposals/blob/main/proposals/revise_use_of_layer_metadata/README.md) related to this that may have an impact on design)
_In Exploration_ - sublayers, left vs right hand, cameras, shaders, animation, etc.

<h3>Let us know!</h3>

- Are the options suitable for your needs?
- Do you have any questions or concerns?
- Is there anything missing or you would do differently?
- Excited not to have to manually rotate/scale as much or maintain internal tools?

Thanks! 
Scott Renaud
[scott.renaud@autodesk.com](mailto:scott.renaud@autodesk.com)

**Comment by daviddeacon** [comment]
Hi,

This is something that's a constant struggle with Maya assets typically using centimetres and Houdini liking unit's in meters. So super excited to see this being worked on.

A extra 4th option could be to apply the Axis/Unit conversion on Import, it would be like the 'Axis/Unit conversion on Import' example but freezing the rotation and scale values.

Also a question about the export, when exporting to a different scale does the actual point data get scaled or is it simply written with different meta data? I would really want the option to scale the actual data vs just updating the meta data.

David.

**Reply by scottrenaud** [reply]
Hey David, 

Would freezing on import be used to keep scale/rotation the same as how it was authored?
If let's say you had a scale of 2 on the object (before import) and on import we convert from meters to cm, so we scale up to 200.  If we froze this it would become 1 (losing the artist's original scale value of 2).  Is that what you would expect, or do you think it needs to be similar to FBX where it performs the unit conversion and then restores the original rotation/scale, essentially making the conversion transparent, so in this example seeing 2 at the end, but it's been scaled?

As for export, yep the idea is to scale the data in addition to updating the metadata.  And actually, we're kind of getting into the same question as above, once rotated/scaled do you think it should also be frozen or have the option?

Thanks for the feedback!
Scott

**Reply by daviddeacon** [reply]
Hi Scott,
That's a good point, all of our assets have clean zero values on there root transform so I didn't consider them to have anything different. I think the example that is similar to FBX makes a lot of sense,  and aligns with a pre existing knowledge of someone using FBX.

This would be the same for export data, so if you export something that's none zero scale, during export it's internally scaled and you get the same values on the transforms on the exported file as you had when exporting.

I'm really looking forward to this feature, when is it likely to hit v0.28.0?

Also are you considering cameras, and scaling their extra attributes, focal distance, clipping planes etc? 

David

**Reply by scottrenaud** [reply]
Hey David,

Alright, good to know and definitely something we'll explore.
There's a load of stuff we would like to support so we'll have to phase things and get feedback along the way.
I can’t say which release it’ll be part of, but we’re making preparations for the work.

Thanks!
Scott

**Comment by kimuss** [comment]
yessssss , finally, goddamn :D 
goodbye houdini's problem
- considering cameras "stuff" would be extremely good too

thank you for your work on it

**Comment by AlessBernardi** [comment]
Hi,
to me, it looks like the option of changing the root node value or adding a new one should best fit my use cases.
The problem is when you want to save your files back, how would you track that the conversion has already been applied and avoid adding another conversion each time you open/save it?

From a pipeline standpoint, I would experiment using expression variables and assume one up vector, tweaking its value, based on the DCC used, ex: build on the fly a root layer that holds a variable for which vector is up, and then sublayer the asset I want to load with the expression on a correction node, this way you should be able to save your edits at the right place, not loosing the expression.
As mentioned, it has not been tested yet.

Thanks
Alex
